### PR TITLE
Shell command archive

### PR DESCRIPTION
**Pull request by jkloetzke**
Refactors the binary artifact handling and provides a "shell" backend that might use any command for the actual up- or download.

### CODE REVIEW

File: pym/bob/archive.py [hunk 1]
```diff
@@ -0,0 +1,389 @@
...
+        (tmpFd, tmpName) = mkstemp()
+        try:
+            os.close(tmpFd)
+            with tarfile.open(tmpName, mode="w:gz") as tar:
+                tar.add(path, arcname=".")
+
+            env = { k:v for (k,v) in os.environ.items() if k in self.__whiteList }
+            env["BOB_LOCAL_ARTIFACT"] = tmpName
+            env["BOB_REMOTE_ARTIFACT"] = self._makeUrl(buildId)
+            ret = subprocess.run(["/bin/bash", "-ec", self.__uploadCmd],
```
Comment (rhubert) [inline pym/bob/archive.py]: Maybe I'm wrong but this limits uploadCmd to a cmd given in default.yaml? I'd prefer to put my commands into a separate shell script and/or have the full string functions. E.g.:
```yaml
archive:
       -
           backend: shell
           upload: "$<<scripts/myscript>> upload ${BOB_LOCAL_ARTIFACT} ${BOB_REMOTE_ARTIFACT}"
           upload: "$<<scripts/myscript>> download ${BOB_LOCAL_ARTIFACT} ${BOB_REMOTE_ARTIFACT}"
```

**Comment by rhubert** [comment]
I'm thinking about changing the behavior for the download mode to not break working recipes through updating bob. Maybe all archives can be used for download unless tagged with:
```yaml
use: [upload, no-download]
```
?

File: pym/bob/archive.py [hunk 2]
```diff
@@ -0,0 +1,389 @@
...
+                                 stdout=subprocess.DEVNULL, cwd="/tmp", env=env)
+            if ret.returncode == 0:
+                removePath(path)
+                os.makedirs(path)
+                with tarfile.open(tmpName, "r:gz", errorlevel=1) as tar:
+                    tar.extractall(path)
+                print(colorize("ok", "32"))
+                success = True
+            else:
+                print(colorize(str(e.reason), "33"))
```
Comment (jkloetzke) [inline pym/bob/archive.py]: Correct. Will fix it...

File: pym/bob/archive.py (hunk 1 above)
Comment (jkloetzke) [inline pym/bob/archive.py]: Interesting idea. On the other hand you can write a whole script directly in default.yaml if you want as block text (like you do e.g. in `buildScript`). I thought about string functions but these commands have no real context. They are just there to transfer tar files to/from a remote server.

**Comment by jkloetzke** [comment]
The default, if no `use` key is given, is `[download, upload]`. This is the same as before where this key did not exist. Do I overlook something? I'm not sure I understood your comment correctly...

**Comment by rhubert** [comment]
I must have overlooked this default. This is what I wanted to say..

File: pym/bob/archive.py (hunk 1 above)
Comment (rhubert) [inline pym/bob/archive.py]: Sure you can, but do you really want to write complex scripts in yaml? 
Maybe for doing more complex thinks like uploading to different servers depending on the package or something like this it's better to have a plugin interface for uploading?

**Comment by jkloetzke** [comment]
Updated to work with Python 3.3 and 3.4 too...

**Pull request merged**